File: Java/RelativityClient/src/main/java/com/nuix/relativityclient/utils/JdbcClient.java
package com.nuix.relativityclient.utils;

import com.nuix.relativityclient.relativitytypes.FieldColumn;
import org.apache.log4j.LogManager;
import org.apache.log4j.Logger;

import java.sql.Date;
import java.sql.*;
import java.util.*;

public class JdbcClient {
    private static final Logger LOGGER = LogManager.getLogger(JdbcClient.class.getName());
    private Connection connection;
    private String databaseName;
    private String fileShareUrl;
    private String dataGridUrl;
    private String serverName;

    public JdbcClient(String sqlServer, Integer port, String instance, String domain, String username, String password) throws SQLException {

        LOGGER.info("Creating JdbcClient");
        String connectionString = "jdbc:jtds:sqlserver://"+sqlServer;
        if (port != null){
            connectionString+=":"+port;
        }

        if (instance != null && instance.length()>0){
            connectionString+=";instance="+instance;
        }

        if (domain != null && domain.length()>0){
            connectionString+=";domain="+domain;
        }
        connection = DriverManager.getConnection(connectionString,username,password);
    }

    public String getFileShareUrl() {
        return fileShareUrl;
    }

    public String getDataGridUrl() {
        return dataGridUrl;
    }

    public String getServerName() {
        return serverName;
    }

    public int getCaseArtifactId() throws SQLException {
        String selectQuery = "SELECT [ArtifactID]" +
                " FROM "+databaseName+".[Artifact]" +
                " WHERE [ArtifactTypeID]=8";

        Statement statement = connection.createStatement();
        ResultSet resultSet = statement.executeQuery(selectQuery);
        while (resultSet.next()) {
            return resultSet.getInt(1);
        }
        throw new SQLException("Could not find case Artifact ID");
    }

    public void setWorkspaceArtifactId(int artifactId) throws SQLException {

        databaseName = "[EDDS"+artifactId+"].[EDDSDBO]";
        LOGGER.info("Set database "+databaseName);
    }

    public void fetchWorkspaceResourceServer(int artifactId) throws SQLException {
        String selectQuery = "SELECT [ServerID]"+
                ",[DefaultFileLocationCodeArtifactID]" +
                ",[DataGridFileShareResourceServerArtifactID]"+
                " FROM [EDDS].[eddsdbo].[Case]" +
                " WHERE [ArtifactID]=?";

        PreparedStatement statement = connection.prepareStatement(selectQuery);
        statement.setInt(1,artifactId);

        ResultSet resultSet = statement.executeQuery();
        while (resultSet.next()) {
            int serverId=resultSet.getInt(1);
            int fileShareId=resultSet.getInt(2);
            int dataGridId=resultSet.getInt(3);

            serverName = getResourceServerProperty(serverId,"Name");
            fileShareUrl= getResourceServerProperty(fileShareId,"URL");
            dataGridUrl = getResourceServerProperty(dataGridId,"URL");
            break;
        }
    }

    private String getResourceServerProperty(int artifactId,String propertyName) throws SQLException {
        String selectQuery = "SELECT ["+propertyName+"]"+
                " FROM [EDDS].[eddsdbo].[ResourceServer]" +
                " WHERE [ArtifactID]=?";

        PreparedStatement statement = connection.prepareStatement(selectQuery);
        statement.setInt(1,artifactId);

        ResultSet resultSet = statement.executeQuery();
        while (resultSet.next()) {
            return resultSet.getString(1);
        }
        return null;
    }

    public Map<String, FieldColumn> getFieldColumns() throws SQLException {
        Map<String, FieldColumn> fieldColumns = new HashMap<>();
        String selectQuery = "SELECT [Field].[ArtifactID]" +    // 1
                ",[Field].[FieldArtifactTypeID]" +              // 2
                ",[Field].[DisplayName]" +                      // 3
                ",[Field].[ArtifactViewFieldID]" +              // 4
                ",[ArtifactViewField].[ColumnName]" +           // 5
                ",[ArtifactViewField].[ItemListType]" +         // 6
                ",[Field].[CodeTypeID]" +                       // 7
                " FROM "+databaseName+".[Field] AS [Field]" +
                " INNER JOIN  "+databaseName+".[ArtifactViewField] as [ArtifactViewField]" +
                " ON [Field].[ArtifactViewFieldID]=[ArtifactViewField].[ArtifactViewFieldID]" +
                " WHERE [Field].[FieldArtifactTypeID] = 10";

        Statement statement = connection.createStatement();
        ResultSet resultSet = statement.executeQuery(selectQuery);
        while (resultSet.next()) {
            int fieldArtifactId = resultSet.getInt(1);
            String fieldName = resultSet.getString(3);
            String columnName= resultSet.getString(5);
            String columnType= resultSet.getString(6);
            int codeTypeId= resultSet.getInt(7);
            FieldColumn fieldColumn;
            if (codeTypeId>0) {
                fieldColumn = new FieldColumn(columnName, columnType, fieldArtifactId, codeTypeId);
            }else {
                fieldColumn = new FieldColumn(columnName, columnType, fieldArtifactId);
            }
            fieldColumns.put(fieldName,fieldColumn);
        }
        return fieldColumns;
    }

    public int getDocumentArtifactId(String docId, FieldColumn identifier) throws SQLException {
        String selectQuery = "SELECT [ArtifactID]" +        // 1
                " FROM " + databaseName + ".[Document]" +
                " WHERE [" + identifier.getColumnName() + "] = ?";

        PreparedStatement statement = connection.prepareStatement(selectQuery);
        statement.setString(1, docId);
        ResultSet resultSet = statement.executeQuery();
        while (resultSet.next()) {
            return resultSet.getInt(1);
        }
        throw new SQLException("Document does not exist in workspace");
    }


    public int createArtifact(int artifactType,int parentArtifactId,int containerId, String textIdentifier) throws SQLException {
        String insertQuery = "INSERT INTO "+databaseName + ".[Artifact]"+
                " ([ArtifactTypeID]"+                // 1
                ",[ParentArtifactID]"+               // 2
                ",[AccessControlListID]"+            // 3
                ",[AccessControlListIsInherited]"+   // 4
                ",[CreatedOn]"+                      // 5
                ",[LastModifiedOn]"+                 // 6
                ",[LastModifiedBy]"+                 // 7
                ",[CreatedBy]"+                      // 8
                ",[ContainerID]"+                    // 9
                ",[Keywords]"+                       // 10
                ",[Notes]"+                          // 11
                ",[DeleteFlag]"+                     // 12
                ",[TextIdentifier])"+                // 13
                " VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?)";

        PreparedStatement statement = connection.prepareStatement(insertQuery,Statement.RETURN_GENERATED_KEYS);

        statement.setInt(1, artifactType);
        statement.setInt(2, parentArtifactId);
        statement.setInt(3, 1);
        statement.setInt(4, 1);
        statement.setDate(5, new Date(System.currentTimeMillis()));
        statement.setDate(6, new Date(System.currentTimeMillis()));
        statement.setInt(7, 777);
        statement.setInt(8, 777);
        statement.setInt(9, containerId);
        statement.setString(10, "");
        statement.setString(11, "");
        statement.setBoolean(12, false);
        statement.setString(13, textIdentifier);
        int affectedRows = statement.executeUpdate();

        if (affectedRows == 0) {
            throw new SQLException("Creating Artifact failed, no rows affected.");
        }

        try (ResultSet generatedKeys = statement.getGeneratedKeys()) {
            if (generatedKeys.next()) {
                return generatedKeys.getInt(1);
            }
            else {
                throw new SQLException("Creating Artifact failed, no ID obtained.");
            }
        }
    }


    public void setDocumentChoiceValues(int documentArtifactId, int fieldArtifactId, int codeTypeId, int containerId, Set<String> values, boolean overlay) throws SQLException {
        if (overlay) {
            String deleteStatement = "DELETE FROM " + databaseName + ".[ZCodeArtifact_" + codeTypeId + "]" +
                    "WHERE [AssociatedArtifactID]=?";
            PreparedStatement deleteAssociationStatement = connection.prepareStatement(deleteStatement);
            deleteAssociationStatement.setInt(1, documentArtifactId);
            deleteAssociationStatement.executeUpdate();
        }

        for (String value : values) {
            String selectQuery = "SELECT [ArtifactID]" +
                    " FROM " + databaseName + ".[Code]" +
                    " WHERE [CodeTypeID]=? AND [Name]=?";

            PreparedStatement selectCodeId = connection.prepareStatement(selectQuery);
            selectCodeId.setInt(1, codeTypeId);
            selectCodeId.setString(2, value);

            ResultSet resultSet = selectCodeId.executeQuery();
            Integer choiceArtifactId = null;
            while (resultSet.next()) {
                choiceArtifactId = resultSet.getInt(1);
            }

            // Track new choice option
            if (choiceArtifactId == null) {
                choiceArtifactId = createArtifact(7, containerId, containerId, value.toString());

                String insertQuery = "INSERT INTO " + databaseName + ".[Code]" +
                        " ([ArtifactID]" +           // 1
                        ",[CodeTypeID]" +            // 2
                        ",[Order]" +                 // 3
                        ",[IsActive]" +              // 4
                        ",[UpdateInSearchEngine]" +  // 5
                        ",[Name])" +                 // 6
                        "VALUES (?, ?, ?, ?, ?, ?)";

                PreparedStatement insertNewChoiceStatement = connection.prepareStatement(insertQuery);
                insertNewChoiceStatement.setInt(1, choiceArtifactId);
                insertNewChoiceStatement.setInt(2, codeTypeId);
                insertNewChoiceStatement.setInt(3, 0);
                insertNewChoiceStatement.setBoolean(4, true);
                insertNewChoiceStatement.setBoolean(5, false);
                insertNewChoiceStatement.setString(6, value);
                insertNewChoiceStatement.executeUpdate();
            }

            // Set choice on document
            String insertQuery = "INSERT INTO " + databaseName + ".[ZCodeArtifact_" + codeTypeId + "]" +
                    " ([CodeArtifactID]" +           // 1
                    ",[AssociatedArtifactID])" +     // 2
                    "VALUES (?, ?)";
            PreparedStatement insertChoiceToDocumentStatement = connection.prepareStatement(insertQuery);
            insertChoiceToDocumentStatement.setInt(1, choiceArtifactId);
            insertChoiceToDocumentStatement.setInt(2, documentArtifactId);
            insertChoiceToDocumentStatement.executeUpdate();
        }
    }

    public Set<String> addDocumentRecord(int artifactId, int parentArtifactId, Map<String, Object> fields, boolean hasNative, String fileIcon, Boolean supportedByViewer, String relativityNativeType) throws SQLException {
        Set<String> incorrectlyFormattedFields = new HashSet<>();
        String insertQuery = "INSERT INTO " + databaseName + ".[Document]" +
                    " ([ArtifactID]" +                   // 1
                    ",[AccessControlListID_D]" +         // 2
                    ",[ParentArtifactID_D]" +            // 3
                    ",[HasNative]" +                     // 4
                    ",[HasAnnotations]" +                // 5
                    ",[HasInlineTags]" +                 // 6
                    ",[FileIcon]" +                      // 7
                    ",[SupportedByViewer]" +             // 8
                    ",[RelativityNativeType]";           // 9

            String valuesSubQuery = "VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?";
            for (String fieldName : fields.keySet()) {
                insertQuery += ",[" + fieldName + "]";
                valuesSubQuery += ", ?";
            }
            insertQuery += ") " + valuesSubQuery + ")";

        PreparedStatement statement = connection.prepareStatement(insertQuery);

        statement.setInt(1, artifactId);
        statement.setInt(2, 1);
        statement.setInt(3, parentArtifactId);
        statement.setBoolean(4, hasNative);
        statement.setBoolean(5, false);
        statement.setBoolean(6, false);
        statement.setString(7, fileIcon);
        statement.setObject(8, supportedByViewer);
        statement.setObject(9, relativityNativeType);

        int fieldPosition=10;

        for (String fieldName : fields.keySet()){
            Object fieldValue = fields.get(fieldName);
            try {
                statement.setObject(fieldPosition, fieldValue);
            } catch (SQLException e1){
                try{
                    statement.setString(fieldPosition, fieldValue.toString());
                } catch (SQLException e2){
                    // Cannot set value as object
                    incorrectlyFormattedFields.add(fieldName);
                    statement.setObject(fieldPosition, null);
                }
            }
            fieldPosition++;
        }

        int affectedRows = statement.executeUpdate();

        if (affectedRows == 0) {
            throw new SQLException("Inserting Document failed, no rows affected.");
        }

        return incorrectlyFormattedFields;
    }

    public List<Integer> getItemAncestors(int artifactId)  throws SQLException {
        String selectQuery = "SELECT [AncestorArtifactID]" +
                " FROM " + databaseName + ".[ArtifactAncestry]" +
                " WHERE [ArtifactID]=?";

        PreparedStatement selectStatement = connection.prepareStatement(selectQuery);
        selectStatement.setInt(1, artifactId);

        ResultSet resultSet = selectStatement.executeQuery();
        List<Integer> ancestors = new ArrayList<>();

        while (resultSet.next()) {
            ancestors.add(resultSet.getInt(1));
        }

        return ancestors;
    }

    public void setAncestors(int artifactId, List<Integer> ancestorIds) throws SQLException {
        String insertQuery = "INSERT INTO " + databaseName + ".[ArtifactAncestry]" +
                " ([ArtifactID]" +                   // 1
                ",[AncestorArtifactID])" +           // 2
                " VALUES (?, ?)";

        PreparedStatement statement = connection.prepareStatement(insertQuery);
        for (int ancestorId : ancestorIds) {
            statement.setInt(1, artifactId);
            statement.setInt(2, ancestorId);
            statement.addBatch();
        }

        statement.executeBatch();
    }


    public void setDataGridFileMapping(int artifactId, int fieldArtifactId, String fileLocation, long fileSize, boolean overlay) throws SQLException {
        if (overlay) {
            String dropQuery = "DELETE FROM " + databaseName + ".[DataGridFileMapping]" +
                    "WHERE [ArtifactID]=? AND [FieldArtifactID]=?";
            PreparedStatement deleteMappingstatement = connection.prepareStatement(dropQuery);
            deleteMappingstatement.setInt(1, artifactId);
            deleteMappingstatement.setInt(2, fieldArtifactId);
            deleteMappingstatement.executeUpdate();
        }

        String insertQuery = "INSERT INTO " + databaseName + ".[DataGridFileMapping]" +
                " ([ArtifactID]" +                  // 1
                ",[FieldArtifactID]" +              // 2
                ",[FileLocation]" +                 // 3
                ",[FileSize])" +                    // 4
                " VALUES (?, ?, ?, ?)";

        PreparedStatement statement = connection.prepareStatement(insertQuery);
        statement.setInt(1, artifactId);
        statement.setInt(2, fieldArtifactId);
        statement.setString(3, fileLocation);
        statement.setLong(4, fileSize);
        int affectedRows = statement.executeUpdate();

        if (affectedRows == 0) {
            throw new SQLException("Setting data grid file mapping failed, no rows affected.");
        }
    }

    public void setFileshareFileMapping(int artifactId, String guid, String filename, String identifier, String fileLocation, long fileSize, boolean overlay) throws SQLException {
        if (overlay) {
            String dropQuery = "DELETE FROM " + databaseName + ".[File]" +
                    "WHERE [DocumentArtifactID]=?";
            PreparedStatement deleteMappingstatement = connection.prepareStatement(dropQuery);
            deleteMappingstatement.setInt(1, artifactId);
            deleteMappingstatement.executeUpdate();
        }

        String insertQuery = "INSERT INTO " + databaseName + ".[File]" +
                " ([Guid]" +                  // 1
                ",[DocumentArtifactID]" +     // 2
                ",[Filename]" +               // 3
                ",[Order]" +                  // 4
                ",[Type]" +                   // 5
                ",[Rotation]" +               // 6
                ",[Identifier]" +             // 7
                ",[Location]" +               // 8
                ",[InRepository]" +           // 9
                ",[Size]" +                   // 10
                ",[Details]" +                // 11
                ",[Billable])" +              // 12
                " VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?)";

        PreparedStatement statement = connection.prepareStatement(insertQuery);
        statement.setString(1, guid);
        statement.setInt(2, artifactId);
        statement.setString(3, filename);
        statement.setInt(4, 0);
        statement.setInt(5, 0);
        statement.setInt(6, -1);
        statement.setString(7, identifier);
        statement.setString(8, fileLocation);
        statement.setBoolean(9, true);
        statement.setLong(10, fileSize);
        statement.setObject(11, null);
        statement.setBoolean(12, true);
        int affectedRows = statement.executeUpdate();

        if (affectedRows == 0) {
            throw new SQLException("Setting fileshare file mapping failed, no rows affected.");
        }
    }



    public Set<String> updateDocumentRecord(int artifactId, Map<String, Object> fields, boolean hasNative) throws SQLException {
        Set<String> incorrectlyFormattedFields = new HashSet<>();
        String updateQuery = "UPDATE " + databaseName + ".[Document] SET" +
                    "[HasNative] = ?";                             // 1
            for (String fieldName : fields.keySet()) {
                updateQuery += ",[" + fieldName + "] = ?";

            }
            updateQuery += "WHERE [ArtifactID]=?";

        PreparedStatement statement = connection.prepareStatement(updateQuery);

        statement.setBoolean(1, hasNative);
        int fieldPosition=2;

        for (String fieldName : fields.keySet()){
            Object fieldValue = fields.get(fieldName);
            try {
                statement.setObject(fieldPosition, fieldValue);
            } catch (SQLException e1){
                try{
                    statement.setString(fieldPosition, fieldValue.toString());
                } catch (SQLException e2){
                    // Cannot set value as object
                    incorrectlyFormattedFields.add(fieldName);
                    statement.setObject(fieldPosition, null);
                }
            }
            fieldPosition++;
        }

        statement.setInt(fieldPosition, artifactId);

                int affectedRows = statement.executeUpdate();

        if (affectedRows == 0) {
            throw new SQLException("Updating Document failed, no rows affected.");
        }

        return incorrectlyFormattedFields;
    }

}

File: Ruby/RelativityJdbcUpload/JdbcClient.rb
require 'set'

class JdbcClient

  def initialize(server_name, server_port, server_instance, server_domain, username, password)
    # Build connection string
    connection_string = "jdbc:jtds:sqlserver://#{server_name}"
    connection_string += ":#{server_port}" unless server_port.nil?
    connection_string += ";instance=#{server_instance}" unless server_instance.nil? || server_instance.length == 0
    connection_string += ";domain=#{server_domain}" unless server_domain.nil? || server_domain.length == 0

    @connection = java.sql.DriverManager.get_connection(connection_string, username, password)
  end

  def server_name
    @server_name
  end

  def data_grid_url
    @data_grid_url
  end

  def file_share_url
    @file_share_url
  end

  def workspace_artifact_id=(artifact_id)
    @database_name = "[EDDS#{artifact_id}].[EDDSDBO]"
  end

  def get_case_artifact_id
    select_query = "SELECT [ArtifactID] FROM #{@database_name}.[Artifact] WHERE [ArtifactTypeId] = 8"

    statement = @connection.create_statement
    result_set = statement.execute_query(select_query)

    if result_set.next
      result_set.get_int(1)
    else
      raise java.sql.SQLException.new("Could not find case Artifact ID")
    end
  end

  def fetch_workspace_resource_server(artifact_id)
    select_query = "SELECT"\
      " [ServerID]"\
      ",[DefaultFileLocationCodeArtifactId]"\
      ",[DataGridFileShareResourceServerArtifactID]"\
      "FROM [EDDS].[eddsdbo].[Case] WHERE [ArtifactID] = ?"

    prepared_statement = @connection.prepare_statement(select_query)
    prepared_statement.set_int(1, artifact_id)

    result_set = prepared_statement.execute_query
    result_set.next

    server_id = result_set.get_int(1)
    file_share_id = result_set.get_int(2)
    data_grid_id = result_set.get_int(3)

    @server_name = get_resource_server_property(server_id, "Name")
    @file_share_url = get_resource_server_property(file_share_id, "URL")
    @data_grid_url = get_resource_server_property(data_grid_id, "URL")
  end

  def get_field_columns
    select_query = "SELECT"\
      " [Field].[ArtifactID]"\
      ",[Field].[FieldArtifactTypeID]"\
      ",[Field].[DisplayName]"\
      ",[Field].[ArtifactViewFieldID]"\
      ",[ArtifactViewField].[ColumnName]"\
      ",[ArtifactViewField].[ItemListType]"\
      ",[Field].[CodeTypeID]"\
      " FROM #{@database_name}.[Field] AS [Field]"\
      " INNER JOIN #{@database_name}.[ArtifactViewField] as [ArtifactViewField]"\
      " ON [Field].[ArtifactViewFieldID] = [ArtifactViewField].[ArtifactViewFieldID]"\
      " WHERE [Field].[FieldArtifactTypeID] = 10"

    statement = @connection.create_statement
    result_set = statement.execute_query(select_query)

    field_columns = {}
    while result_set.next do
      field_artifact_id = result_set.get_int(1)
      field_name = result_set.get_string(3)
      column_name = result_set.get_string(5)
      column_type = result_set.get_string(6)
      code_type_id = result_set.get_int(7)

      field_columns[field_name] = {
          :column_name => column_name,
          :column_type => column_type,
          :field_artifact_id => field_artifact_id,
          :code_type_id => code_type_id > 0 ? code_type_id : nil
      }
    end

    field_columns
  end

  def get_document_artifact_id(doc_id, identifier)
    select_query = "SELECT [ArtifactID] FROM #{@database_name}.[Document] WHERE [#{identifier[:column_name]}] = ?"

    prepared_statement = @connection.prepare_statement(select_query)
    prepared_statement.set_string(1, doc_id)

    result_set = prepared_statement.execute_query
    if result_set.next
      result_set.get_int(1)
    else
      raise java.sql.SQLException.new("Document does not exist in workspace")
    end
  end

  def create_artifact(artifact_type, parent_artifact_id, container_id, text_identifier)
    insert_query = "INSERT INTO #{@database_name}.[Artifact]"\
      " ([ArtifactTypeID]"\
      ",[ParentArtifactID]"\
      ",[AccessControlListID]"\
      ",[AccessControlListIsInherited]"\
      ",[CreatedOn]"\
      ",[LastModifiedOn]"\
      ",[LastModifiedBy]"\
      ",[CreatedBy]"\
      ",[ContainerID]"\
      ",[Keywords]"\
      ",[Notes]"\
      ",[DeleteFlag]"\
      ",[TextIdentifier])"\
      " VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?)"

    prepared_statement = @connection.prepare_statement(insert_query, java.sql.Statement.RETURN_GENERATED_KEYS)
    prepared_statement.set_int(1, artifact_type)
    prepared_statement.set_int(2, parent_artifact_id)
    prepared_statement.set_int(3, 1)
    prepared_statement.set_int(4, 1)
    prepared_statement.set_date(5, java.sql.Date.new(Time.now.to_f * 1000))
    prepared_statement.set_date(6, java.sql.Date.new(Time.now.to_f * 1000))
    prepared_statement.set_int(7, 777)
    prepared_statement.set_int(8, 777)
    prepared_statement.set_int(9, container_id)
    prepared_statement.set_string(10, "")
    prepared_statement.set_string(11, "")
    prepared_statement.set_boolean(12, false)
    prepared_statement.set_string(13, text_identifier)

    affected_rows = prepared_statement.execute_update
    if affected_rows == 0
      raise java.sql.SQLException.new("Creating Artifact failed, no rows affected")
    end

    generated_keys = prepared_statement.get_generated_keys

    if generated_keys.next
      generated_keys.get_int(1)
    else
      raise java.sql.SQLException.new("Creating Artifact failed, no ID obtained")
    end
  end

  def set_document_choice_values(document_artifact_id, code_type_id, container_id, values, overlay)
    if overlay
      delete_statement = "DELETE FROM #{@database_name}.[ZCodeArtifact_#{code_type_id}] WHERE [AssociatedArtifactID] = ?"

      delete_association_statement = @connection.prepare_statement(delete_statement)
      delete_association_statement.set_int(1, document_artifact_id)
      delete_association_statement.execute_update
    end

    values.each do |value|
      select_query = "SELECT [ArtifactID] FROM #{@database_name}.[Code] WHERE [CodeTypeID] = ? AND [Name] = ?"

      select_code_id = @connection.prepare_statement(select_query)
      select_code_id.set_int(1, code_type_id)
      select_code_id.set_string(2, value)

      result_set = select_code_id.execute_query
      choice_artifact_id = nil
      while result_set.next do
        choice_artifact_id = result_set.get_int(1)
      end

      # Track new choice option
      if choice_artifact_id.nil?
        choice_artifact_id = create_artifact(7, container_id, container_id, value)

        choice_insert_query = "INSERT INTO #{@database_name}.[Code]"\
          " ([ArtifactID]"\
          ",[CodeTypeID]"\
          ",[Order]"\
          ",[IsActive]"\
          ",[UpdateInSearchEngine]"\
          ",[Name])"\
          " VALUES (?, ?, ?, ?, ?, ?)"

        insert_new_choice_statement = @connection.prepare_statement(choice_insert_query)
        insert_new_choice_statement.set_int(1, choice_artifact_id)
        insert_new_choice_statement.set_int(2, code_type_id)
        insert_new_choice_statement.set_int(3, 0)
        insert_new_choice_statement.set_boolean(4, true)
        insert_new_choice_statement.set_boolean(5, false)
        insert_new_choice_statement.set_string(6, value)
        insert_new_choice_statement.execute_update
      end

      # Set choice on document
      insert_query = "INSERT INTO #{@database_name}.[ZCodeArtifact_#{code_type_id}]"\
        " ([CodeArtifactID]"\
        ",[AssociatedArtifactID])"\
        " VALUES (?, ?)"

      insert_choice_to_document_statement = @connection.prepare_statement(insert_query)
      insert_choice_to_document_statement.set_int(1, choice_artifact_id)
      insert_choice_to_document_statement.set_int(2, document_artifact_id)
      insert_choice_to_document_statement.execute_update
    end
  end

  def add_document_record(artifact_id, parent_artifact_id, fields, has_native, file_icon, supported_by_viewer, relativity_native_type)
    insert_query = "INSERT INTO #{@database_name}.[Document]"\
      " ([ArtifactID]"\
      ",[AccessControlListID_D]"\
      ",[ParentArtifactID_D]"\
      ",[HasNative]"\
      ",[HasAnnotations]"\
      ",[HasInlineTags]"\
      ",[FileIcon]"\
      ",[SupportedByViewer]"\
      ",[RelativityNativeType]"

    values_sub_query = "VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?"
    fields.each_key do |fieldName|
      insert_query << ",[#{fieldName}]"
      values_sub_query << ", ?"
    end

    insert_query << ") #{values_sub_query})"

    prepared_statement = @connection.prepare_statement(insert_query)
    prepared_statement.set_int(1, artifact_id)
    prepared_statement.set_int(2, 1)
    prepared_statement.set_int(3, parent_artifact_id)
    prepared_statement.set_boolean(4, has_native)
    prepared_statement.set_boolean(5, false)
    prepared_statement.set_boolean(6, false)
    prepared_statement.set_string(7, file_icon)
    prepared_statement.set_object(8, supported_by_viewer)
    prepared_statement.set_object(9, relativity_native_type)

    field_position = 10
    incorrectly_formatted_fields = Set.new

    fields.each do |fieldName, fieldValue|
      begin
        prepared_statement.set_object(field_position, fieldValue)
      rescue java.sql.SQLException
        begin
          prepared_statement.set_string(field_position, fieldValue.to_s)
        rescue java.sql.SQLException
          # Cannot set value as object
          incorrectly_formatted_fields.add(fieldName)
          prepared_statement.set_object(field_position, nil)
        end
      end
      field_position += 1
    end

    affected_rows = prepared_statement.execute_update
    if affected_rows == 0
      raise java.sql.SQLException("Inserting Document failed, no rows affected")
    end

    incorrectly_formatted_fields
  end

  def get_item_ancestors(artifact_id)
    select_query = "SELECT [AncestorArtifactID] FROM #{@database_name}.[ArtifactAncestry] WHERE [ArtifactID] = ?"

    prepared_statement = @connection.prepare_statement(select_query)
    prepared_statement.setInt(1, artifact_id)

    result_set = prepared_statement.execute_query

    ancestors = []
    while result_set.next
      ancestors << result_set.get_int(1)
    end

    ancestors
  end

  def set_ancestors(artifact_id, ancestor_ids)
    insert_query = "INSERT INTO #{@database_name}.[ArtifactAncestry]"\
      " ([ArtifactID]"\
      ",[AncestorArtifactID])"\
      " VALUES (?, ?)"

    prepared_statement = @connection.prepareStatement(insert_query)
    ancestor_ids.each do |ancestor_id|
      prepared_statement.set_int(1, artifact_id)
      prepared_statement.set_int(2, ancestor_id)

      prepared_statement.add_batch
    end

    prepared_statement.execute_batch
  end

  def set_data_grid_file_mapping(artifact_id, field_artifact_id, file_location, file_size, overlay)
    if overlay
      drop_query = "DELETE FROM #{@database_name}.[DataGridFileMapping] WHERE [ArtifactID] = ? AND [FieldArtifactID] = ?"

      drop_statement = @connection.prepare_statement(drop_query)
      drop_statement.set_int(1, artifact_id)
      drop_statement.set_int(2, field_artifact_id)

      drop_statement.execute_update
    end

    insert_query = "INSERT INTO #{@database_name}.[DataGridFileMapping]"\
      " ([ArtifactID]"\
      ",[FieldArtifactID]"\
      ",[FileLocation]"\
      ",[FileSize])"\
      " VALUES (?, ?, ?, ?)"

    insert_statement = @connection.prepare_statement(insert_query)
    insert_statement.set_int(1, artifact_id)
    insert_statement.set_int(2, field_artifact_id)
    insert_statement.set_string(3, file_location)
    insert_statement.set_long(4, file_size)

    affected_rows = insert_statement.execute_update
    if affected_rows == 0
      raise java.sql.SQLException.new("Setting data grid file mapping failed, no rows affected")
    end
  end

  def set_file_share_file_mapping(artifact_id, guid, filename, identifier, file_location, file_size, overlay)
    if overlay
      drop_query = "DELETE FROM #{@database_name}.[File] WHERE [DocumentArtifactID] = ?"

      drop_statement = @connection.prepare_statement(drop_query)
      drop_statement.set_int(1, artifact_id)

      drop_statement.execute_update
    end

    insert_query = "INSERT INTO #{@database_name}.[File]"\
      " ([Guid]"\
      ",[DocumentArtifactID]"\
      ",[Filename]"\
      ",[Order]"\
      ",[Type]"\
      ",[Rotation]"\
      ",[Identifier]"\
      ",[Location]"\
      ",[InRepository]"\
      ",[Size]"\
      ",[Details]"\
      ",[Billable]"\
      " VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?)"

    insert_statement = @connection.prepare_statement(insert_query)
    insert_statement.set_string(1, guid)
    insert_statement.set_int(2, artifact_id)
    insert_statement.set_string(3, filename)
    insert_statement.set_int(4, 0)
    insert_statement.set_int(5, 0)
    insert_statement.set_int(6, -1)
    insert_statement.set_string(7, identifier)
    insert_statement.set_string(8, file_location)
    insert_statement.set_boolean(9, true)
    insert_statement.set_long(10, file_size)
    insert_statement.set_object(11, nil)
    insert_statement.set_boolean(12, true)

    affected_rows = insert_statement.execute_update
    if affected_rows == 0
      raise java.SQLException.new("Setting fileshare file mapping failed, no rows affected")
    end
  end

  def update_document_record(artifact_id, fields, has_native)
    update_query = "UPDATE #{@database_name}.[Document] SET [HasNative] = ?"

    fields.each_key do |fieldName|
      update_query << ",[#{fieldName}] = ?"
    end
    update_query << " WHERE [ArtifactID] = ?"

    update_statement = @connection.prepare_statement(update_query)
    update_statement.set_boolean(1, has_native)

    field_position = 2
    incorrectly_formatted_fields = Set.new

    fields.each do |fieldName, fieldValue|
      begin
        update_statement.set_object(field_position, fieldValue)
      rescue java.sql.SQLException
        begin
          update_statement.set_string(field_position, fieldValue)
        rescue java.sql.SQLException
          # Cannot set value as object
          incorrectly_formatted_fields.add(fieldName)
          update_statement.set_object(field_position, nil)
        end
      end
      field_position += 1
    end

    update_statement.set_int(field_position, artifact_id)

    affected_rows = update_statement.execute_update
    if affected_rows == 0
      raise java.sql.SQLException.new("Updating Document failed, no rows affected")
    end

   incorrectly_formatted_fields
  end

  private def get_resource_server_property(artifact_id, property_name)
    select_query = "SELECT [#{property_name}] FROM [EDDS].[eddsdbo].[ResourceServer] WHERE [ArtifactId] = ?"

    prepared_statement = @connection.prepare_statement(select_query)
    prepared_statement.set_int(1, artifact_id)

    result_set = prepared_statement.execute_query
    result_set.next ? result_set.get_string(1) : nil
  end

end